[Product Query] Add support for the Filter By Price Block

gigitux opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR adds support for the Filter By Price block. 

The code is written to be easily extendible and adding easily the support to other filters and other variations (frontend side/editor side).


<!-- Reference any related issues or PRs here -->

#6790

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### Screenshots

<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->
![Screen Capture on 2022-09-15 at 14-26-59](https://user-images.githubusercontent.com/4463174/190422830-81f428c5-c80e-41f3-a18d-115c36836531.gif)


### Testing

#### Automated Tests


#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Create a new post/page and add the Product Query block and the Filter By Price block.
2. On the front-end side, apply some range of price and be sure that the filters are applied correctly.
3. Edit the post/page and add the Product Query block and enable "Show only products on sale" option.
4. On the front-end side, apply some range of price and be sure that the filters are applied correctly.
5. Edit the post/page and add a Query Loop block.
6. On the front-end side, apply some range of price, be sure that the filters are applied correctly and the Query Loop works correctly (continue to show posts).
7. Do the same steps (1-6), but instead add in a post/page, add the blocks in an FSE template (for example Product Catalog template).

> **Note**
> During your test, pay attention to the pagination too.


* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> N/A

gigitux committed:
Product Query: Fix pagination issue

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-update

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/price_filter_support

gigitux committed:
Product Query - Add support for the Filter By Price Block #6790

Product Query - Add support for the Filter By Price Block

Aljullu commented on src/BlockTypes/ProductQuery.php:
```diff
@@ -79,24 +82,39 @@ public function get_query_by_attributes( $query ) {
 			'post_status'    => 'publish',
 			'posts_per_page' => $query['posts_per_page'],
 			'orderby'        => $query['orderby'],
+			'orderby'        => $query['orderby'],
```
This line is duplicated, is that a typo?

Aljullu commented on src/BlockTypes/ProductQuery.php:
```diff
@@ -117,4 +135,71 @@ private function get_on_sale_products_query( $variation_props ) {
+	 * @return array
+	 */
+	private function get_queries_by_attributes( $variation_props ) {
+		return array(
+			'on_sale' => ( ! isset( $variation_props['attributes']['query']['onSale'] ) || true !== $variation_props['attributes']['query']['onSale'] ) ? array() : $this->get_on_sale_products_query(),
+		);
```
I know it's not introduced in this PR, but I wonder if we could extract this check into its own variable to make the code easier to read, something like:

```suggestion
		$on_sale = ! isset( $variation_props['attributes']['query']['onSale'] ) || true !== $variation_props['attributes']['query']['onSale'];
		return array(
			'on_sale' => $on_sale ? array() : $this->get_on_sale_products_query(),
		);
```

Aljullu reviewed (commented):
That's really cool. :star_struck: Really exciting to see that this works in parallel to the _On sale_ filter, the potential is huge! Good job! :clap: 

I found a strange behavior. When no products match the price filter, all products are shown (ie, setting the price filter to 33-34):

![imatge](https://user-images.githubusercontent.com/3616980/190647601-02b19aac-513a-4a6c-bde8-10c3620979c1.png)

I also left a couple of inline comments, but they are quite minor.

Oh, and can you set the sprint of this PR? :pray:

gigitux committed:
fix query relation

gigitux commented:
Thanks for the review and for testing the PR. I addressed all your feedback! [b243528](https://github.com/woocommerce/woocommerce-blocks/pull/7162/commits/b243528f2dc703917063a352e1bfe4cb8c8a2545)

The relation in the query was wrong. It was `OR`, but the right value is `AND`.

I noticed this strange behavior. 
With this range `33-34` and when the `Product Query` block shows all products, the result matches with WC Core

![image](https://user-images.githubusercontent.com/4463174/190685908-c00dfa78-53e1-49fc-acb7-2ef95c3dd1c1.png)

*above WC Core*

With the same range, but with the `Product Query` block that shows only on-sale products, the result doesn't match 

![image](https://user-images.githubusercontent.com/4463174/190686531-71f5328a-c45a-4638-88f8-d5721262336d.png)

### Exploration

It seems that it isn't easy to find the `grouped products on sale` with the `WP_Query` class. With [e55a41d](https://github.com/woocommerce/woocommerce-blocks/pull/7162/commits/e55a41dcf2f3bd366edb7837b1d8d8eca0c39c1b) I have tried a different approach, and it works! The problem that I see, is that we do two queries to the database, but it seems that the function 
[`wc_get_product_ids_on_sale()`]( https://hooks.wbcomdesigns.com/reference/functions/wc_get_product_ids_on_sale/ ) caches the results, so, in the end, it could be more performant.

What do you think?

gigitux committed:
fix on sale query

gigitux committed:
Merge branch 'trunk' into add/price_filter_support

gigitux committed:
Merge branch 'trunk' into add/price_filter_support

Aljullu commented on src/BlockTypes/ProductQuery.php:
```diff
@@ -80,41 +83,107 @@ public function get_query_by_attributes( $query ) {
-					'type'    => 'numeric',
-				),
+				'relation' => 'AND',
+				$max_price_query,
+				$min_price_query,
```
Should we check that `$max_price_query` and `$min_price_query` are not an empty array, here? Otherwise we end up with queries in this form:

```
    [meta_query] => Array
        (
            [relation] => AND
            [0] => Array
                (
                )

            [1] => Array
                (
                )

        )
```

(I have no idea if that's an issue, though, so feel free to ignore this comment if you think that's ok)

gigitux replied:
I added the check in this PR https://github.com/woocommerce/woocommerce-blocks/pull/7186/files#diff-efb947e345af5e501dfaa2980b055b3b31c308b31e8db23049b6b48fecee33d8R227-R228

Aljullu commented on src/BlockTypes/ProductQuery.php:
```diff
@@ -80,41 +83,107 @@ public function get_query_by_attributes( $query ) {
+	 */
+	private function get_on_sale_products_query() {
+		return array(
+			'post_in' => wc_get_product_ids_on_sale(),
```
Should this be [`post__in`](https://developer.wordpress.org/reference/classes/wp_query/#post-page-parameters) (with two underscores), instead?

Aljullu commented on src/BlockTypes/ProductQuery.php:
```diff
@@ -80,41 +83,107 @@ public function get_query_by_attributes( $query ) {
+				$queries_attributes,
+				$queries_filters
+			),
+			function( $acc, $query ) {
+						// Ignoring the warning of not using meta queries.
+						// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_query
+						$acc['meta_query'] = isset( $query['meta_query'] ) ? array_merge( $acc['meta_query'], array( $query['meta_query'] ) ) : $acc['meta_query'];
+						return $acc;
+			},
```
If we go with the `post__in` approach, this `array_reduce()` is ignoring that property, right? Because we are only taking the `meta_query` value from `$queries_attributes` and `$queries_filters`, but not `post__in`. I might be missing something, though. :thinking:

Aljullu reviewed (commented):
Thanks for updating this PR, @gigitux! I'm facing some issues with the last changes, though. I left some inline comments below. :point_down:

gigitux committed:
fix bugged pagination and on-sale filter after refactor

gigitux committed:
Merge branch 'add/price_filter_support' of https://github.com/woocommerce/woocommerce-blocks into add/price_filter_support

gigitux commented:
With the last refactor I broke:
- pagination
- on filter sale

Thanks for your review and comments! Now, it should be fine! 🤞

gigitux committed:
fix reload page when the Filter By Price block is used with All Products block

gigitux committed:
Merge branch 'trunk' into add/price_filter_support

Aljullu commented on src/BlockTypes/ProductQuery.php:
```diff
@@ -80,41 +83,111 @@ public function get_query_by_attributes( $query ) {
+				$queries_filters
+			),
+			function( $acc, $query ) {
+				if ( isset( $query['post__in'] ) ) {
+					$acc['post__in'] = isset( $acc['post__in'] ) ? array_merge( $acc['post__in'], $query['post__in'] ) : $query['post__in'];
+				}
```
If I'm understanding this correctly, I think this is working now because we only set the `post__in` attribute with the _on sale_ filter. But if at some point there were two filters using `post__in`, instead of concatenating the ids we should only include the intersection of `$acc['post__in']` and `$query['post__in']`, in other words, ids which are in both arrays so it behaves as an AND instead of an OR.

gigitux replied:
Good point! I fixed this!

Aljullu reviewed (approved):
Thanks for updating this PR, @gigitux! This is working fine. I left one minor comment about how we merge the `post__in` attribute, instead of accumulating all ids I think we should only accumulate the duplicated ones. But besides this, LGTM so pre-approving.

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/price_filter_support

gigitux committed:
use array_intersect instead of array_merge

gigitux commented:
Thanks for the review! 🥳
